aaelf64-morello/aaelf64-morello.rst

```diff
@@ -865,3 +929,195 @@ veneer is used. The BX changes the execution state from A64 to C64:
+and bounds such that the entire TLS static block is accessible.
+
+Resolver functions
+^^^^^^^^^^^^^^^^^^
+
+A resolver function takes arguments in c0 (address of the TLS GOT slot), and c2
+(a copy of the thread pointer) and returns a pointer to the TLS global in c0.
+The resolver function has a custom calling convention that must preserve all
+registers except c0 and c1.
+
+Considerations:
+
+- Any dynamically loaded modules will be placed outside of the bounds of the
+  thread pointer, so a resolver function cannot return an offset from the
+  thread pointer, but rather needs to return a pointer (capability).
+- To minimize reading of ``CTPIDR_EL0``, the resolver functions take a
+  copy of ``CTPIDR_EL0`` as an argument and preserve it.
+
+
+Static TLS block resolver
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If the TLS variable is in the static block, while resolving the
+``R_MORELLO_TLSDESC`` relocation, the dynamic linker will place in the two GOT
+slots associated with this variable:
+
+- A capability to the static TLS block resolver function at offset 0.
+
+- The offset of the variable in the static TLS block at offset 16 (8 bytes).
+
+- The size of the variable at offset 24 (8 bytes).
+
+An implementation of the static block resolver could be the following:
+
+.. code-block:: text
+
+  ldp x0, x1, [c0, #16]
+  add c0, c2, x0
+  scbnds c0, c0, x1
+  ret c30
+
+Local Exec
+^^^^^^^^^^
+
+The capability to the TLS variable is derived from ``CTPIDR_EL0``. There
+are no requirements on how this is performed or the registers used, except
+that the sequence doesn't produce a dynamic relocation. A possible
+instruction sequence could be:
+
+.. code-block:: text
+
+  mrs c0, CTPIDR_EL0
+  movz x8, #:tprel_g1:local_exec_var
+  movk x8, #:tprel_g0_nc:local_exec_var
+  movz x9, #:size_g1:local_exec_var
+  movk x9, #:size_g0_nc:local_exec_var
+  add c0, c0, x8, uxtx
+  scbnds c0, c0, x9
+
+Initial Exec
+^^^^^^^^^^^^
+
+The capability to the TLS variable is derived from ``CTPIDR_EL0``. The size
+and offset of the TLS variable is stored in a GOT slot (first 8 bytes contains the
+offset and the second 8 bytes the size). This GOT slot is initialized by a
+``R_MORELLO_TPREL128`` dynamic relocation. The access must use the
+``R_MORELLO_TLSIE_ADR_GOTTPREL_PAGE20`` and ``R_MORELLO_TLSIE_ADD_LO12``
+relocations in order to allow relaxation to Local Exec. There are no other
+requirements on how this is performed or the registers used. A possible
+instruction sequence could be:
+
+.. code-block:: text
+
+  adrp c0, :gottprel:initial_exec_var
+  add c0, c0, :gottprel_lo12:initial_exec_var
+  ldp x0, x8, [c0]
+  mrs c1, CTPIDR_EL0
+  add c0, c1, x0, uxtx
+  scbnds c0, c0, x8
+
+Initial Exec to Local Exec relaxation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The linker will generate 16 bytes in the ``.rodata`` section, containing
+the offset in the static TLS block in the first 8 bytes and the size of the
+symbol in the next 8 bytes:
```

Comment:
W.r.t. requiring that the linker has generated 16 bytes in the .rodata section.
As it corresponds to strange linker scripts which put .rodata input sections in a differently named output section:
Is this an input section name or output section name?

Also, is there a particular need to specify this?
I'd imagine any section marked read-only would work (though I can't say I know for certain).